libbpf-tools/bashreadline.bpf.c

```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "bashreadline.h"
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+SEC("uretprobe/readline")
+int printret(struct pt_regs *ctx) {
```

Comment:
try using BPF_KRETPROBE() macro, see other tools

Reply:
So, even it is an uprobe?

Reply:
yeah, uprobe and kprobe are no different in terms of input context, so BPF_KPROBE/BPF_KRETPROBE can be used for uprobe/uretprobe